### ISSUE TITLE
Feature roadmap

### ISSUE THREAD

**Issue by sdispater** [python-poetry/poetry#1856]
This is the official feature roadmap for Poetry. This gives an overview of future features and ideas that will be implemented in Poetry and in which release they will be available.

Each feature release will have its own Github project to track the progress being made on these new features

This roadmap is subject to change and features will be added as needed.

## 1.1
- [x] Lightweight core library (#1993).
- [x] First phase of improving the packages installation speed : Download distributions separately before installing them (#2595).
- [x] Opt-in option to disable the generation of the `setup.py` file.
- [x] Drop Python 3.4 support. Note that Poetry will still be able to manage Python 3.4 environments.
- [x] Python 2.7 and 3.5 deprecation (#2683).

## 1.2
- [x] Plugin system (#693, #1237).
- [ ] Implementation of a `bundle` command, see #537 for more context. This command should be pluggable via the plugin system (#1992, #2682).
- [ ] Improve dependency resolution. Better support for system specific wheels and speed improvements.
- [ ] Development of a package registry to speed up dependency resolution of packages with missing metadata.
- [ ] Second phase of improving the packages installation speed: Implementation of a separate library to replace calls to `pip` for installation. This will be an opt-in feature in this release.
- [x] Add support for dependency groups (#1644). 
- [x] Drop Python 2.7 and 3.5 support.
## 1.3
- [ ] Third phase of improving the package installation speed : Make the phase 2 feature the default one with the possibility to opt out.

## 1.4
- [ ] Fourth phase of improving the package installation speed : Disable the possibility to opt out of the  phase 2 feature as it should be stable by now.

**Comment by marctorrellas** [comment]
Something I would like to see is the ability to use different channels. In conda you can specify e.g.

```
name: dslib
channels:
  - anaconda
  - conda-forge
dependencies:
  # python related packages
  - python==3.6.5
  - scikit-learn 
```

However, you cannot install anything outside of conda, which is really bad. That's why (among other things) I and other people I know use poetry. 

On the other hand, poetry only supports pip, which is not perfect as explained here https://www.anaconda.com/understanding-conda-and-pip/

TLDR:

- There is never a need to have compilers available to install them. You download binaries directly, so installation is cleaner and faster
- Additionally conda packages are not limited to Python software.

Is there any chance to get support for conda? I would like to see something like channels, and poetry checking first conda, then pip. This would probably improve installation speed as well

What do you reckon? (I'm not an full-time dev, so please let me know if something stated above is wrong)

**Comment by etijskens** [comment]
I fully support this request. It is very reasonable and will serve many users.
Kindest regards,

Dr [Engel]bert Tijskens - HPC Analyst/Consultant
Flemish Supercomputer Center
University of Antwerp / Computational mathematics
++32 3 265 3879, ++32 494 664408

On 12 Jan 2020, at 14:16, Marc Torrellas <notifications@github.com<mailto:notifications@github.com>> wrote:


Something I would like to see is the ability to use different channels. In conda you can specify e.g.

name: dslib
channels:
  - anaconda
  - conda-forge
dependencies:
  # python related packages
  - python==3.6.5
  - scikit-learn


However, you cannot install anything outside of conda, which is really bad. That's why (among other things) I and other people I know use poetry.

On the other hand, poetry only supports pip, which is not perfect as explained here https://www.anaconda.com/understanding-conda-and-pip/<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fwww.anaconda.com%2Funderstanding-conda-and-pip%2F&data=02%7C01%7Cengelbert.tijskens%40uantwerpen.be%7C9cb0fdc766c8425dd63a08d7976198c3%7C792e08fb2d544a8eaf72202548136ef6%7C0%7C0%7C637144317741297240&sdata=ihwfbq5eH0NHrxYSi7AZz9HqFriSnPiF5s5fJFhssPY%3D&reserved=0>

TLDR:

  *   There is never a need to have compilers available to install them. You download binaries directly, so installation is cleaner and faster
  *   Additionally conda packages are not limited to Python software.

Is there any chance to get support for conda? I would like to see something like channels, and poetry checking first conda, then pip. This would probably improve installation speed as well

What do you reckon? (I'm not an full-time dev, so please let me know if something stated above is wrong)

**Comment by sdispater** [comment]
Thanks for your interest in Poetry and for your feedback.

However, this issue is not the place to make feature requests but is reserved for discussion about the current roadmap detailed above. If you want to make a feature request, please open a proper issue with the appropriate template.

That being said there is no plan to add support for conda at the moment.

**Comment by Mattwmaster58** [comment]
I'm curious about these 1.1 roadmap items:
 - Improve dependency resolution. Better support for system specific wheels and speed improvements.
 - First phase of improving the packages installation speed: Download distributions separately before installing them.

Do we have hard numbers on how long % wise each of these steps take? I'm not saying we have to, but it would be interesting to note. My guess would be on package installation speed.

**Comment by seansfkelley** [comment]
Is something along the lines of "refine the CLI's UX" suitable for the roadmap? Most of the internals of Poetry I'm quite happy with (well, except for the speed issues already covered), but I find myself cobbling together workflows from commands that aren't _quite_ what I need, but can be abused to make it so.

**Comment by jtratner** [comment]
Small note: Python 3.4 is already EOL as of March 2019, so if it's causing you trouble, seems reasonable to drop support for it - see Python announcement here - https://www.python.org/downloads/release/python-3410/

**Comment by sdispater** [comment]
@jtratner Yes, we will drop support for Python 3.4 in the next feature release without going through a deprecation plan.

This decision was taken because, as mentioned above, it was cumbersome and it seems that people have already moved on to more recent versions (see https://pypistats.org/packages/poetry)

I'll update the roadmap to reflect that.

**Comment by pietrodn** [comment]
A prioritary issue to be included in the next releases is to fix the hash checking of dependencies, IMHO.
See: https://github.com/python-poetry/poetry/issues/2422#issuecomment-822540374

**Comment by pietrodn** [comment]
Would it be possible to cut a patch release to include the fix for #2422, which is a security issue? 🙏

**Comment by adriangb** [comment]
Would it be possible to move the roadmap to a quarterly roadmap and the implement a consistent release cadence? It's kind of a shame that some implemented, working features are only available in pre releases (presumably because they're waiting on other features that are supposed to be included in the same minor release)

**Comment by thedrow** [comment]
I think we can check #1644 as done.

On another note, could we please release this feature and re-plan the roadmap? This one is really useful and I'd like to be able to use it.

**Comment by thedrow** [comment]
If there's a need, I'm willing to cut 1.2 myself and release it to the public.
Having dependency groups will improve the workflow for all our users.

**Comment by s0undt3ch** [comment]
👍 for a release with dependency groups.

**Comment by finswimmer** [comment]
> If there's a need, I'm willing to cut 1.2 myself and release it to the public.

How would this be different from the current alpha release?

**Comment by thedrow** [comment]
It would allow us to use this feature in production in conjunction with [nox-poetry](https://github.com/cjolowicz/nox-poetry) for example. Pre-versions are usually not allowed in production packages.

The feature itself is functional and nothing is preventing its release.
I'd like to be able to install this version without `--pre` or without working around dependency issues for packages which depend on poetry.

**Comment by jakeberesford-palmetto** [comment]
Releasing 1.2 with groups ahead of other features slated for 1.2 would help my team hugely, we would prefer not to use a pre-release in production.

**Comment by emekdahl-palmetto** [comment]
Bump! We need this!

**Comment by adriangb** [comment]
> How would this be different from the current alpha release?

I've encountered bugs/issues that have been fixed since the last pre-release that forced me to abandon the pre-release and hop onto the last stable. Granted, this could be fixed with a new pre-release, but that hasn't happened in a while, which discourages me from using any pre-release.

**Comment by mitom** [comment]
It would be very good to get a clear answer to the prompt for an early 1.2 release specifically for the groups feature (even if the answer is no). Currently, I am stuck in a limbo as this feature would be one of very few things that I need from a dependency manager, and knowing if it has a chance to land in weeks or potentially several months would be a key part in deciding whether we stick with poetry or not.

On a more practical side, it looks like the groups feature is one of 3 non-transparent changes planned for 1.2 (correct me if I misinterpret it, these changes being plugins, groups, bundle command) and the rest are speed improvements. With an external eye, we could have the groups functionality without the rest, so why not?

A 3rd note - if we could have a pre-release with just the groups - assuming that will avoid the issues mentioned above with pre-releases, it could solve this need for at least a portion of us (even though I understand not all of us). Would this be an option?

**Comment by haf** [comment]
Ref https://github.com/python-poetry/poetry/discussions/5459

**Comment by g-as** [comment]
> Would love to know timelines for a 1.2 release. Plugins (specifically the bundling plugin) is important for my use case

See #5586

**Comment by opserx** [comment]
I think peotry is perfect except for the speed of package installation(or update).  Why not focus on fixing speed in a minor version?  Of course, I can't thank you enough for your contributions and respect your roadmap.

**Comment by neersighted** [comment]
Closing this for now as we don't have a definite roadmap -- but this is long out of date and https://github.com/python-poetry/roadmap is the current (also dusty) location of our planning
